### PR TITLE
Fix NullPointerException when using plugin dependencies in version 1.6.0

### LINKED ISSUE #76: NullPointerException with inlcudePluginDependecies=true
Using mvn 3.3.3
Minimal test case POM (obviously the used dependencies are bogus, but it should be enough to replicate.

`<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>
<groupId>test</groupId>
  <artifactId>PluginDependencies</artifactId>
  <version>0.1-SNAPSHOT</version>  
  <build>
    <pluginManagement>
      <plugins>
        <plugin>
          <groupId>org.codehaus.mojo</groupId>
          <artifactId>exec-maven-plugin</artifactId>
          <version>1.6.0</version>
          <executions>
            <execution>
              <id>default</id>
              <goals>
                <goal>java</goal>
              </goals>
            </execution>
          </executions>
          <configuration>
            <executableDependency>
              <groupId>org.codehaus.mojo</groupId>
              <artifactId>exec-maven-plugin</artifactId>
            </executableDependency>
            <includeProjectDependencies>false</includeProjectDependencies>
            <includePluginDependencies>true</includePluginDependencies>
            <mainClass>org.codehaus.mojo.exec.ExecJavaMojo</mainClass>
          </configuration>
          <dependencies>
            <dependency>
              <groupId>org.codehaus.mojo</groupId>
              <artifactId>exec-maven-plugin</artifactId>
              <version>1.6.0</version>
              <type>jar</type>
            </dependency>
          </dependencies>
        </plugin>
      </plugins>
    </pluginManagement>
  </build>
</project>`

This will fail with
` mvn exec:java
[INFO] Scanning for projects...
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building PluginDependencies 0.1-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- exec-maven-plugin:1.6.0:java (default-cli) @ PluginDependencies ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 0.673 s
[INFO] Finished at: 2017-03-16T11:46:02+01:00
[INFO] Final Memory: 11M/309M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.6.0:java (default-cli) on project PluginDependencies: Execution default-cli of goal org.codehaus.mojo:exec-maven-plugin:1.6.0:java failed. NullPointerException -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
Picked up _JAVA_OPTIONS: -Xmx2048M

in 1.5.0 this setup worked fine. Obviously with this sample POM the build will still fail because the executableDependency has no mainClass.

### PR DESCRIPTION
**Pull request by jonasrutishauser**
Fixes #76 

A regression was introduced with the following change:
exec mojo can run executableDependency instead of executable (7684096)

### CODE REVIEW

**Comment by masu1402** [comment]
When is a new version of the API with this fix to be expected?

**Comment by rfscholte** [comment]
There are conflicts with this PR, so you need to update your patch first. Be aware that we don't like protected fields. Also, this might already be solved with #122 maybe you can verify that first.

**Comment by jonasrutishauser** [comment]
I've updated the patch as it is still needed.
The protected field is now replaced with a protected getter and an `IllegalArgumentException` introduced with #122 is fixed too.

**Comment by lwaggonerExpedia** [comment]
@jonasrutishauser @rfscholte - I just verified this change locally in a separate project, the change is required because the pluginDependencies field is duplicated. If we reference pluginDependencies in AbstractExecMojo it will always be null and the direct casts to hashmaps create null pointers. Leaving the field private means we need a get method in AbstractExecMojo otherwise we can't get the pluginDependencies in ExecJavaMojo. This alone gets you further but the filter can't be null and that is the last change. It would be interesting to understand why this wasn't reproduced in a sample project/test, I didn't investigate that far.

**Comment by lwaggonerExpedia** [comment]
I've found an additional edge case and bug: in the case where this is used in a multi module project

This returns a different result:
MavenProject executableProject =
                    this.projectBuilder.build(executablePomArtifact, buildingRequest).getProject()

so if you execute from the sub pom it works, if you execute from the root pom your executable will be missing and it will fail all the time.

I'll put a note in the PR

**Comment by starksm64** [comment]
Three changes from this PR allowed me to build the plugin locally and use with the includePluginDependencies setting.

```diff
diff --git a/src/main/java/org/codehaus/mojo/exec/AbstractExecMojo.java b/src/main/java/org/codehaus/mojo/exec/AbstractExecMojo.java
index bd926e1..fdbe0ae 100644
--- a/src/main/java/org/codehaus/mojo/exec/AbstractExecMojo.java
+++ b/src/main/java/org/codehaus/mojo/exec/AbstractExecMojo.java
@@ -262,6 +262,11 @@ public abstract class AbstractExecMojo
         return session;
     }
 
+    protected final List<Artifact> getPluginDependencies()
+    {
+        return pluginDependencies;
+    }
+
     /**
      * Examine the plugin dependencies to find the executable artifact.
      * 
diff --git a/src/main/java/org/codehaus/mojo/exec/ExecJavaMojo.java b/src/main/java/org/codehaus/mojo/exec/ExecJavaMojo.java
index 7b47684..b6ace87 100644
--- a/src/main/java/org/codehaus/mojo/exec/ExecJavaMojo.java
+++ b/src/main/java/org/codehaus/mojo/exec/ExecJavaMojo.java
@@ -28,6 +28,8 @@ import org.apache.maven.plugins.annotations.ResolutionScope;
 import org.apache.maven.project.MavenProject;
 import org.apache.maven.project.ProjectBuilder;
 import org.apache.maven.project.ProjectBuildingRequest;
+import org.apache.maven.shared.artifact.filter.resolve.AndFilter;
+import org.apache.maven.shared.artifact.filter.resolve.TransformableFilter;
 import org.apache.maven.shared.transfer.artifact.resolve.ArtifactResult;
 import org.apache.maven.shared.transfer.dependencies.DefaultDependableCoordinate;
 import org.apache.maven.shared.transfer.dependencies.resolve.DependencyResolver;
@@ -52,12 +54,6 @@ public class ExecJavaMojo
     @Component
     private ProjectBuilder projectBuilder;
 
-    /**
-     * @since 1.1-beta-1
-     */
-    @Parameter( readonly = true, defaultValue = "${plugin.artifacts}" )
-    private List<Artifact> pluginDependencies;
-
     /**
      * The main class to execute.<br>
      * With Java 9 and above you can prefix it with the modulename, e.g. <code>com.greetings/com.greetings.Main</code>
@@ -625,7 +621,7 @@ public class ExecJavaMojo
             if ( this.executableDependency == null )
             {
                 getLog().debug( "All Plugin Dependencies will be included." );
-                relevantDependencies = new HashSet<Artifact>( this.pluginDependencies );
+                relevantDependencies = new HashSet<Artifact>( getPluginDependencies() );
             }
             else
             {
@@ -661,7 +657,9 @@ public class ExecJavaMojo
             MavenProject executableProject =
                 this.projectBuilder.build( executablePomArtifact, buildingRequest ).getProject();
 
-            for ( ArtifactResult artifactResult : dependencyResolver.resolveDependencies( buildingRequest, executableProject.getModel(), null ) )
+            for ( ArtifactResult artifactResult : dependencyResolver.resolveDependencies( buildingRequest,
+                                                                                          executableProject.getModel(),
+                                                                                          new AndFilter(Collections.<TransformableFilter>emptyList() ) ) )
             {
                 executableDependencies.add( artifactResult.getArtifact() );
             }
```

**Comment by slawekjaranowski** [comment]
Thanks for PR and for patience 😄
